### PR TITLE
Fix crash in libunwind while interpreting debug info

### LINKED ISSUE #60460: libunwind: getSavedRegister: invalid read from memory
@alexey-milovidov Are you OK with us re-opening this please?

We're seeing the same thing on all of our production clusters running v23.3, but it's re-producible in clean builds from master from my testing.

What we see:
ClickHouse servers across our entire fleet are occasionally segfaulting, with only the below information in logs:
```
2024.02.24 13:21:53.198661 [ 7 ] {} <Fatal> Application: Child process was terminated by signal 11
```
No consistent pattern in the query log or server logs  stand out to me before or after we see the segfaults.

Reviewing coredumps, here is an example stack trace (almost identical to a few on this ticket): 

```
gdb) bt
#0  0x0000000015da2fd1 in libunwind::DwarfInstructions<libunwind::LocalAddressSpace, libunwind::Registers_x86_64>::getSavedRegister (addressSpace=..., registers=..., cfa=cfa@entry=140650633216840, savedReg=...) at ./contrib/libunwind/src/DwarfInstructions.hpp:111
#1  0x0000000015da1f81 in libunwind::DwarfInstructions<libunwind::LocalAddressSpace, libunwind::Registers_x86_64>::stepWithDwarf (addressSpace=..., pc=<optimized out>, fdeStart=<optimized out>, registers=..., isSignalFrame=@0x7fedb2ae2609: true) at ./contrib/libunwind/src/DwarfInstructions.hpp:258
#2  0x0000000015da1aa2 in libunwind::UnwindCursor<libunwind::LocalAddressSpace, libunwind::Registers_x86_64>::stepWithDwarfFDE (this=0x7fedb2ae2508) at ./contrib/libunwind/src/UnwindCursor.hpp:1002
#3  libunwind::UnwindCursor<libunwind::LocalAddressSpace, libunwind::Registers_x86_64>::step (this=0x7fedb2ae2508) at ./contrib/libunwind/src/UnwindCursor.hpp:2818
#4  0x0000000015da11c8 in unw_backtrace (buffer=0x7fedb2ae2718, size=45) at ./contrib/libunwind/src/libunwind.cpp:350
#5  0x000000000ec18088 in StackTrace::tryCapture (this=0x7fedb2ae2708) at ./src/Common/StackTrace.cpp:287
#6  StackTrace::StackTrace (this=0x7fedb2ae2708, signal_context=...) at ./src/Common/StackTrace.cpp:258
#7  0x000000000edffead in signalHandler (sig=<optimized out>, info=0x7fedb2ae2bf0, context=0x7fedb2ae2ac0) at ./src/Daemon/BaseDaemon.cpp:148
#8  <signal handler called>
#9  0x0000000015da2fd1 in libunwind::DwarfInstructions<libunwind::LocalAddressSpace, libunwind::Registers_x86_64>::getSavedRegister (addressSpace=..., registers=..., cfa=cfa@entry=140650633216840, savedReg=...) at ./contrib/libunwind/src/DwarfInstructions.hpp:111
#10 0x0000000015da1f81 in libunwind::DwarfInstructions<libunwind::LocalAddressSpace, libunwind::Registers_x86_64>::stepWithDwarf (addressSpace=..., pc=<optimized out>, fdeStart=<optimized out>, registers=..., isSignalFrame=@0x7febc7077799: true) at ./contrib/libunwind/src/DwarfInstructions.hpp:258
#11 0x0000000015da1aa2 in libunwind::UnwindCursor<libunwind::LocalAddressSpace, libunwind::Registers_x86_64>::stepWithDwarfFDE (this=0x7febc7077698) at ./contrib/libunwind/src/UnwindCursor.hpp:1002
#12 libunwind::UnwindCursor<libunwind::LocalAddressSpace, libunwind::Registers_x86_64>::step (this=0x7febc7077698) at ./contrib/libunwind/src/UnwindCursor.hpp:2818
#13 0x0000000015da11c8 in unw_backtrace (buffer=0x7febc7077838, size=45) at ./contrib/libunwind/src/libunwind.cpp:350
#14 0x000000000ec18088 in StackTrace::tryCapture (this=0x7febc7077828) at ./src/Common/StackTrace.cpp:287
#15 StackTrace::StackTrace (this=0x7febc7077828, signal_context=...) at ./src/Common/StackTrace.cpp:258
#16 0x000000000ec2c5cb in DB::(anonymous namespace)::writeTraceInfo (trace_type=DB::TraceType::Real, info=<optimized out>, context=0x7febc7077d80) at ./src/Common/QueryProfiler.cpp:68
#17 <signal handler called>
#18 0x0000000015ce6629 in LZ4_compress_default (src=<optimized out>, dst=<optimized out>, srcSize=<optimized out>, maxOutputSize=<optimized out>) at ./contrib/lz4/lib/lz4.c:1387
#19 0x0000000012ea3a53 in DB::ICompressionCodec::compress (this=0x7fe0d527b620, source=0x7fd72fab4040 "\003", source_size=65536, dest=0x7fd72db49abb "\202e151024b\"\003") at ./src/Compression/ICompressionCodec.cpp:88
#20 0x0000000012e36d00 in DB::CompressedWriteBuffer::nextImpl (this=0x7fed6b6d4cd8) at ./src/Compression/CompressedWriteBuffer.cpp:42
#21 0x0000000014353684 in std::__1::__uninitialized_allocator_copy[abi:v15000]<std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::regex_token_iterator<std::__1::__wrap_iter<char const*>, char, std::__1::regex_traits<char> >, std::__1::regex_token_iterator<std::__1::__wrap_iter<char const*>, char, std::__1::regex_traits<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*>(std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >&, std::__1::regex_token_iterator<std::__1::__wrap_iter<char const*>, char, std::__1::regex_traits<char> >, std::__1::regex_token_iterator<std::__1::__wrap_iter<char const*>, char, std::__1::regex_traits<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*) (__alloc=..., __first1=..., __last1=..., __first2=0x606f100 <vtable for DB::HashingWriteBuffer+16>) at ./contrib/llvm-project/libcxx/include/__memory/uninitialized_algorithms.h:545
#22 0x0000000000000000 in ?? ()
```

Seeing the below frame:
```
#16 0x000000000ec2c5cb in DB::(anonymous namespace)::writeTraceInfo (trace_type=DB::TraceType::Real, info=<optimized out>, context=0x7febc7077d80) at ./src/Common/QueryProfiler.cpp:68
```

It seems like our case aligns with what @azat mentioned about this occuring as a result of query profiling, so we can try and turn them off as a first step just to confirm.

These frames are also pretty consistent from the coredumps I have reviewed so far:
```
#18 0x0000000015ce6629 in LZ4_compress_default (src=<optimized out>, dst=<optimized out>, srcSize=<optimized out>, maxOutputSize=<optimized out>) at ./contrib/lz4/lib/lz4.c:1387
#19 0x0000000012ea3a53 in DB::ICompressionCodec::compress (this=0x7fe0d527b620, source=0x7fd72fab4040 "\003", source_size=65536, dest=0x7fd72db49abb "\202e151024b\"\003") at ./src/Compression/ICompressionCodec.cpp:88
#20 0x0000000012e36d00 in DB::CompressedWriteBuffer::nextImpl (this=0x7fed6b6d4cd8) at ./src/Compression/CompressedWriteBuffer.cpp:42
```
Could this be an issue of trying to unwind a stack trace of an INSERT query that is behaving unexpectedly?

Another pattern that I noticed is that we see far more segfaults on ClickHouse replicas that receive fairly heavy INSERT workloads, but I will review some more coredumps and investigate a bit more to make sure this is definitely the case.

You're welcome to assign it to me if you want to and I will try to dig a bit deeper and find the cause if we don't already know what it is.

_Originally posted by @seandhaynes in https://github.com/ClickHouse/ClickHouse/issues/33531#issuecomment-1966611482_

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix a crash in asynchronous stack unwinding (such as when using the sampling query profiler) while interpreting debug info. This closes #60460.

Interesting if this will help.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 676519ce90a5cf3e39415885ad72c09d5202b09a with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60468/676519ce90a5cf3e39415885ad72c09d5202b09a/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60468/676519ce90a5cf3e39415885ad72c09d5202b09a/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Daemon/BaseDaemon.cpp [hunk 1]
```diff
@@ -137,10 +137,18 @@ static void terminateRequestedSignalHandler(int sig, siginfo_t *, void *)
 
 static std::atomic_flag fatal_error_printed;
 
+/// Special handling for a function isPointerReadable.
+thread_local bool checking_pointer = false;
+thread_local jmp_buf signal_jump_buffer;
```
Comment (al13n321) [inline src/Daemon/BaseDaemon.cpp:142]:
```suggestion
thread_local sigjmp_buf signal_jump_buffer;
```

**Review by al13n321 — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@azat, this is just a report from gdb about a signal. After this change, we allow the segmentation of fault signals at runtime when they are produced inside libunwind.

**Comment by azat** [comment]
FWIW this PR makes unwidning at least 2x faster (due to removing `mincore` I guess), the test involves too much extra stuff, but still it shows the improvement.

<details>

### Preparation

```
cat > cat config.d/overrides.yml <<<EOL
logger:
  level: none
EOL

root@bb17d67109a1:~/.server# clickhouse-server -- --trace_log=
Processing configuration file 'config.xml'.
There is no file 'config.xml', will use embedded config.
Merging configuration file 'config.d/overrides.yml'.
```

```
chc -q 'truncate table system.trace_log'; time clickhouse-benchmark -t 30 -c 300 -q 'select sleep(3)' --query_profiler_cpu_time_period_ns=0 --query_profiler_real_time_period_ns=1; chc -nm -q 'system flush logs; select count() from system.trace_log;'
```

Below I will mention `period`, this is sampling period - https://github.com/ClickHouse/ClickHouse/blob/6f74bebb488f669dbfd19dd662d7bb78b735d838/src/Common/QueryProfiler.cpp#L155

### upstream

#### `period=1000000`
```
Stopping launch of queries. Requested time limit 30 seconds is exhausted.
Queries executed: 2929.
localhost:9000, queries: 2929, QPS: 88.673, RPS: 88.673, MiB/s: 0.000, result RPS: 88.673, result MiB/s: 0.000.

0.000%          3.002 sec.
10.000%         3.009 sec.
20.000%         3.012 sec.
30.000%         3.015 sec.
40.000%         3.018 sec.
50.000%         3.023 sec.
60.000%         3.030 sec.
70.000%         3.043 sec.
80.000%         3.069 sec.
90.000%         3.286 sec.
95.000%         3.617 sec.
99.000%         5.013 sec.
99.900%         6.153 sec.
99.990%         6.477 sec.



real    0m33.106s
user    0m7.724s
sys     0m2.787s
1'331'387
```
#### `period=100000`
```
Stopping launch of queries. Requested time limit 30 seconds is exhausted.
Queries executed: 451.
localhost:9000, queries: 451, QPS: 12.055, RPS: 12.055, MiB/s: 0.000, result RPS: 12.055, result MiB/s: 0.000.

0.000%          3.058 sec.
10.000%         3.399 sec.
20.000%         7.653 sec.
30.000%         11.116 sec.
40.000%         17.415 sec.
50.000%         24.347 sec.
60.000%         27.344 sec.
70.000%         29.608 sec.
80.000%         31.865 sec.
90.000%         33.944 sec.
95.000%         35.551 sec.
99.000%         36.496 sec.
99.900%         36.995 sec.
99.990%         36.995 sec.



real    0m37.487s
user    0m3.245s
sys     0m1.738s
951637

Stopping launch of queries. Requested time limit 30 seconds is exhausted.
Queries executed: 541.
localhost:9000, queries: 541, QPS: 14.101, RPS: 14.101, MiB/s: 0.000, result RPS: 14.101, result MiB/s: 0.000.

0.000%          3.078 sec.
10.000%         3.553 sec.
20.000%         5.609 sec.
30.000%         7.059 sec.
40.000%         9.604 sec.
50.000%         14.985 sec.
60.000%         25.467 sec.
70.000%         27.233 sec.
80.000%         29.867 sec.
90.000%         34.375 sec.
95.000%         35.447 sec.
99.000%         37.696 sec.
99.900%         37.812 sec.
99.990%         37.909 sec.



real    0m38.445s
user    0m3.552s
sys     0m1.705s
1005665
```
### [SIGSEGV via signal](https://github.com/ClickHouse/ClickHouse/pull/60468)

#### `period=100000`
```
Stopping launch of queries. Requested time limit 30 seconds is exhausted.
Queries executed: 2290.
localhost:9000, queries: 2290, QPS: 68.366, RPS: 68.366, MiB/s: 0.000, result RPS: 68.366, result MiB/s: 0.000.

0.000%          3.117 sec.
10.000%         3.298 sec.
20.000%         3.353 sec.
30.000%         3.407 sec.
40.000%         3.455 sec.
50.000%         3.499 sec.
60.000%         3.564 sec.
70.000%         3.651 sec.
80.000%         3.900 sec.
90.000%         4.625 sec.
95.000%         5.276 sec.
99.000%         6.288 sec.
99.900%         8.210 sec.
99.990%         8.317 sec.



real    0m33.570s
user    0m4.267s
sys     0m2.093s
2333543
```

</details>

I guess it is not merged yet, because it crashes CI? Maybe we should remap `SIGSEGV` into a different signal i.e. `SIGBUS`, inside clickhouse, and force gdb to ignore `SIGSEGV`, that way the CI will not crash, since this patch is really needed. @alexey-milovidov thoughts?

P.S. I've been looking into the original issue, tried **lots** of things, have lots of details but so far don't understood is it possible to fix it somehow properly.

**Comment by azat** [comment]
> Maybe we should remap SIGSEGV into a different signal i.e. SIGBUS, inside clickhouse, and force gdb to ignore SIGSEGV, that way the CI will not crash, since this patch is really needed.

@alexey-milovidov Or we can simply rebase after https://github.com/ClickHouse/ClickHouse/pull/63865, and the problem with CI should be gone.

**Comment by azat** [comment]
Although maybe mincore (or any quirk for reading invalid addresses) will be still needed, if the original problem was due to wrong DWARF, but will see

**Comment by alexey-milovidov** [comment]
FYI this pull request is primarily for speed - because `mincore` is extremely slow in some environments.
It was also an attempt to generalize the ways to protect is from crashes, but it no longer seems relevant.

**Comment by CheSema** [comment]
This commit breaks the Stateless tests!
Look at this: https://s3.amazonaws.com/clickhouse-test-reports/61047/07e746d5a4ef7f858872a622d4e13b38ac24133a/stateless_tests__aarch64_.html

Sigsev has happen in libunwind, all the program stopped by gdb. Gdb collected the traces and after that program just continues to executing. But for the time while it has been paused, some tests got timeouts.

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/pull/65247

**Pull request merged**